### PR TITLE
dissect-image: support images without rootfs but with /usr partition + support simple partition versioning via strverscmp() on part label

### PR DESCRIPTION
**Pull request by poettering**
This beefs up the dissection a little bit:

1. support is added for images that lack a rootfs but contain a /usr partition, i.e. fully stateless images, that come up every time as pristine. This means nspawn will automatically handle these images and mount a tmpfs a rootfs. Quite similar to what nspawn's --volatile=yes does, but with this option directly built into the image, instead of requiring this to be requested explicitly.

2. if a multiple root fs (or /usr fs) are included in a single image: let's tweak how we determine which one to mount when dissecting the image. Previously we'd let the first one always win. With this change we'll pick the "newest" (as decided by strverscmp() on the partition label). This allows a simply A/B style versioning logic in images: boot one partition, update the other, and use the partition label to indicate which one i newer.

### CODE REVIEW

**Comment by poettering** [comment]
THis is new stuff, please don't hurry thi into v248. This is v249 material

**Comment by bluca** [comment]
don't we want to support DISSECT_IMAGE_USR_NO_ROOT for RootImage in src/core/namespace.c and portabled too?

File: src/shared/dissect-image.c [hunk 1]
```diff
@@ -456,6 +456,22 @@ static int device_wait_for_initialization_harder(
 
 #define DEVICE_TIMEOUT_USEC (45 * USEC_PER_SEC)
 
+static void dissected_partition_done(DissectedPartition *p) {
+        assert(p);
+
+        free(p->fstype);
```
Comment (bluca) [inline src/shared/dissect-image.c:462]: given the pointer is not invalidated and numeric values are reset to -1 below, wouldn't mfree to NULL these be safer?

File: src/shared/dissect-image.c [hunk 2]
```diff
@@ -506,7 +506,9 @@ int dissect_image(
         /* Probes a disk image, and returns information about what it found in *ret.
          *
          * Returns -ENOPKG if no suitable partition table or file system could be found.
-         * Returns -EADDRNOTAVAIL if a root hash was specified but no matching root/verity partitions found. */
+         * Returns -EADDRNOTAVAIL if a root hash was specified but no matching root/verity partitions found.
+         * Returns -ENXIO if we couldn't find any partition suitable as root or /usr partition
+         * Returns -ENOTUNIQ if we only found multiple generic partitions and thus dont know what to do with that */
```
Comment (bluca) [inline src/shared/dissect-image.c]: /dont/don't/

File: src/dissect/dissect.c [hunk 3]
```diff
@@ -44,7 +44,13 @@ static const char *arg_image = NULL;
 static const char *arg_path = NULL;
 static const char *arg_source = NULL;
 static const char *arg_target = NULL;
-static DissectImageFlags arg_flags = DISSECT_IMAGE_REQUIRE_ROOT|DISSECT_IMAGE_DISCARD_ON_LOOP|DISSECT_IMAGE_RELAX_VAR_CHECK|DISSECT_IMAGE_FSCK|DISSECT_IMAGE_USR_NO_ROOT;
+static DissectImageFlags arg_flags =
+        DISSECT_IMAGE_GENERIC_ROOT |
+        DISSECT_IMAGE_REQUIRE_ROOT |
+        DISSECT_IMAGE_DISCARD_ON_LOOP |
```
Comment (bluca) [inline src/dissect/dissect.c]: commit message says:

> There are cases where one wants to use one without the other (i.e. when
inspecting things with systemd-dissect tool it should be OK to do so
even if image has no root fs), and it's cleaner anyway.

But here in systemd-dissect DISSECT_IMAGE_REQUIRE_ROOT is set?

Comment (bluca) [reply src/dissect/dissect.c]: ah nvm, fixed by a following commit

**Review by bluca — APPROVED** [review]
Looks great! Needs a rebase, two minor fixes and an open question

**Comment by DaanDeMeyer** [comment]
From the discoverable partitions spec:

> An operating system should automatically discover and mount the first root partition that does not have the no-auto flag set (as described above) by scanning the disk containing the currently used EFI ESP. It should automatically discover and mount the first /usr/, /home/, /srv/, /var/, /var/tmp/ and swap partitions that do not have the no-auto flag set by scanning the disk containing the discovered root partition.

Should this be relaxed or reworded if we switch systemd to use strverscmp instead of using the first partition found?

File: src/shared/dissect-image.c (hunk 1 above)
Comment (poettering) [reply src/shared/dissect-image.c:462]: C gives the guarantee that on structured initialization unlisted fields are reset to zero. Thus, the initialization will set all those pointers implicitly to NULL for us.

Our general rule hence is to never specify fields during structured initialization that shall be zero/NULL/false anyway.

And thus changing this specific line to `p->fstype = mfree(p->fstype)` is something we could do, but unnecessary hence.

Comment (bluca) [reply src/shared/dissect-image.c:462]: Ah yes of course, there's the structured init later - duh!

**Comment by poettering** [comment]
> don't we want to support DISSECT_IMAGE_USR_NO_ROOT for RootImage in src/core/namespace.c and portabled too?

yes, we do. will fix.

**Comment by poettering** [comment]
> Should this be relaxed or reworded if we switch systemd to use strverscmp instead of using the first partition found?

yes, will fix add.

**Comment by poettering** [comment]
Force pushed a new version, with everything suggested addressed

**Comment by poettering** [comment]
A related PR on the mkosi side that generates images suitable for this is https://github.com/systemd/mkosi/pull/676

**Comment by bluca** [comment]
Given there's the new `label` key=value pair in the json output of systemd-dissect, the test needs to be updated:

https://github.com/systemd/systemd/blob/main/test/units/testsuite-50.sh#L33
https://github.com/systemd/systemd/blob/main/test/units/testsuite-50.sh#L39
https://github.com/systemd/systemd/blob/main/test/units/testsuite-50.sh#L130
https://github.com/systemd/systemd/blob/main/test/units/testsuite-50.sh#L131

**Comment by poettering** [comment]
(hmm, one of those days we should change the test to use some json parser (maybe just jq) to verify the output...)

**Comment by poettering** [comment]
Fixed the CI, putting green label up again

**Pull request merged**